File: backjoon/백트래킹/Python/N과M(5).py
def solve(N, K, array):
  
    result = []

    def dfs():
        if(len(result) == K) :
            print(' '.join(map(str, result)))
            return
        
        for i in range(N):#result에 array의 요소가 없다면 넣어라
            if array[i] not in result:
                result.append(array[i])
                dfs()
                result.pop()

    dfs()

# 메인 함수
if __name__ == "__main__":
    N, K = map(int, input().split())
    array = sorted(list(map(int, input().split())))
    
    solve(N, K, array)

File: backjoon/백트래킹/Js/N과M(5).js
const input = require("fs")
  .readFileSync(process.platform === "linux" ? "./dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const [N, M] = input.shift().split(" ").map(Number);
const array = input[0]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);
let result = "";

function solution(n, m) {
  const seq = [...Array(m)].fill(0); //[0]
  const visited = [...Array(n)].fill(false); //[false, false, false, false]
  const dfs = (k, start) => {
    if (k === m) {
      const arr = [];
      for (let i = 0; i < m; i++) {
        arr.push(seq[i]);
      }
      return (result += `${arr.join(" ")}\n`);
    }

    for (let i = 0; i < array.length; i++) {
      if (!visited[i]) {
        seq[k] = array[i];
        visited[i] = true;
        dfs(k + 1, array[i + 1]);
        visited[i] = false;
      }
    }
  };

  dfs(0, array[0]); //0부터 호출해야함
  return result;
}

console.log(solution(N, M));
